### PR TITLE
Adding option for updating election

### PR DESCRIPTION
**Pull request by dkrusteva1**
Adding functionality for editing elective
HTTP GET to /electives.php/1 returns the form for the elective with id=1 with already filled values.
HTTP POST to /electives.php/1 changes the elective with id=1 if the information is valid
HTTP POST to /electives.php gives the opportunity to enter a new elective

### CODE REVIEW

File: db_task2/dbOperation.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+function getElectiveByID($electiveID){
+    $conn=connection();
+    $stmt = $conn->prepare("SELECT * FROM electives WHERE id=:idPlaceholder limit 1;");
+    $stmt->bindParam(':idPlaceholder', $electiveID);
+    $stmt->execute();
+    return $stmt->fetch(PDO::FETCH_ASSOC);
```
Comment (tantanitaa) [inline db_task2/dbOperation.php]: You can surround it in a try/catch block just in case an error occurs.

File: db_task2/dbOperation.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    $conn=connection();
+    $courseTitle = $_POST['title'];
+    $name = $_POST['name'];
+    $description = $_POST['description'];
+    $stmt=$conn->prepare("UPDATE electives SET title=:titlePlaceholder, description=:descriptionPlaceholder, lecturer=:lecturerPlaceholder WHERE id=:idPlaceholder;");
+    $stmt->bindParam(':idPlaceholder', $electiveID);
+    $stmt->bindParam(':titlePlaceholder', $courseTitle);
+    $stmt->bindParam(':descriptionPlaceholder', $description);
+    $stmt->bindParam(':lecturerPlaceholder', $name);
+    $stmt->execute();
```
Comment (tantanitaa) [inline db_task2/dbOperation.php]: You can surround it in a try/catch block just in case an error occurs.

File: db_task2/dbOperation.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+function addElective (){
+    $conn=connection();
+    $courseTitle = $_POST['title'];
+    $name = $_POST['name'];
+    $description = $_POST['description'];
+    $stmt = $conn->prepare("INSERT INTO electives (title, description, lecturer) VALUES (:courseTitle, :description, :name);");
+    $stmt->bindParam(':courseTitle', $courseTitle);
+    $stmt->bindParam(':description', $description);
+    $stmt->bindParam(':name', $name);
+    $stmt->execute();
```
Comment (tantanitaa) [inline db_task2/dbOperation.php]: You can surround it in a try/catch block just in case an error occurs.

File: db_task2/dbOperation.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+function connection(){
+    $host="localhost";
+    $db="www";
+    $user="root";
+    $password="";
+    $conn  = new PDO("mysql:host=$host; dbname=$db", $user, $password);
+    return $conn;
```
Comment (tantanitaa) [inline db_task2/dbOperation.php]: You can surround it in a try/catch block just in case an error occurs.

File: db_task2/electives.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    validate('name', MAX_LENGTH_LECTURER_NAME, $valid, $errors);
+    validate('description', MAX_LENGTH_DESCRIPTION, $valid, $errors);
+    
+   if (empty($checkIfExists)){
+    if (count($errors)==0){
+        addElective();
+        echo "You have successfully added new elective!";
+       }else{
+           echo "Enter valid information, please!";
+       }
```
Comment (tantanitaa) [inline db_task2/electives.php]: Fix identation, please.

File: db_task2/electives.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+if (!preg_match($regex, $name) && !preg_match($fileName, $name))
+{
+    die("Enter valid path param");
+}
+$electiveDictionary = getElectiveByID($name);
+$checkIfExists=$electiveDictionary["title"];
+if (!$checkIfExists)
+{
+    echo "Please enter existing ID!";
+}
```
Comment (tantanitaa) [inline db_task2/electives.php]: Maybe you can check whether the request method here is get.

**Comment by tantanitaa** [comment]
Good job! :)

File: db_task2/dbOperation.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        $conn=connection();
+        $courseTitle = $_POST['title'];
+        $name = $_POST['name'];
+        $description = $_POST['description'];
+        $stmt=$conn->prepare("UPDATE electives SET title=:titlePlaceholder, description=:descriptionPlaceholder, lecturer=:lecturerPlaceholder WHERE id=:idPlaceholder;");
+        $stmt->bindParam(':idPlaceholder', $electiveID);
+        $stmt->bindParam(':titlePlaceholder', $courseTitle);
+        $stmt->bindParam(':descriptionPlaceholder', $description);
+        $stmt->bindParam(':lecturerPlaceholder', $name);
+        $stmt->execute();
```
Comment (martinely95) [inline db_task2/dbOperation.php:23]: You may want to proactively close the sql connections when you finish with them.
https://www.w3schools.com/php/php_mysql_connect.asp

https://stackoverflow.com/a/15444763/3837638

>it's not pointless to close a connection. Every connection occupies resources in the database server. Most servers are configured with a maximum number of concurrent connections. If you close connections proactively, you free up that slot for another thread, and you get overall better throughput.

This applies to all places where you create a new connection and use it.

File: db_task2/dbOperation.php [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        $name = $_POST['name'];
+        $description = $_POST['description'];
+        $stmt = $conn->prepare("INSERT INTO electives (title, description, lecturer) VALUES (:courseTitle, :description, :name);");
+        $stmt->bindParam(':courseTitle', $courseTitle);
+        $stmt->bindParam(':description', $description);
+        $stmt->bindParam(':name', $name);
+        $stmt->execute();
+    }
+    catch(PDOException $error) {
+        die($error->getMessage());
```
Comment (martinely95) [inline db_task2/dbOperation.php]: Here you are basically outputting to the user the raw message of the exception. Sometimes raw messages may contain sensitive information that we may not want to share with the end users.

Comment (martinely95) [reply db_task2/dbOperation.php]: You may want to change it everywhere.

File: db_task2/electives.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+$regex='/[0-9]+/';
+$name=basename($_SERVER['REQUEST_URI']);
+$fileName='/electives.php/';
+if (!preg_match($regex, $name) && !preg_match($fileName, $name))
+{
+    die("Enter valid path param");
+}
+$electiveDictionary = getElectiveByID($name);
+$checkIfExists=$electiveDictionary["title"];
```
Comment (martinely95) [inline db_task2/electives.php]: `$checkIfExists` - can you think of a better name of the variable? Something that describes what the variable holds?

**Pull request merged**